### ISSUE TITLE
New UI Feedback & Improvements

### ISSUE THREAD

**Issue by hille721** [ansible-community/ara#229]
Hi @dmsimard,

just checked your last work on the UI (PRs #226 and #227):

As I already told you a few times, the new theme is great, the dark one, but also the bright one. Here just some feedback and possible improvements from my side:

* for me it is really hard to read the labels in the dark theme
![image](https://user-images.githubusercontent.com/38468371/112195559-65c67a80-8c0a-11eb-9887-9b30557a5676.png)

* due to the sorting option, the Date and Duration column is not aligned in the playbook index page:
![image](https://user-images.githubusercontent.com/38468371/112196260-22b8d700-8c0b-11eb-9e70-afdcc771783d.png)

* with https://github.com/ansible-community/ara/commit/90550b32393d16a758457ef776ac2fa9e09cc6bf you made the playbook reports card scrollable and constrained the height, which is great! Becasue the three cards are on the same row and the hosts cards is expanded by default, it would also make sense to expand the Files and Records cards per default as well (currently they are collapsed). At least if the screen is wide enough.

* the heading of the file in the file page (e.g. https://dev.demo.recordsansible.org/files/201724.html) is really huge:
![image](https://user-images.githubusercontent.com/38468371/112198534-7cba9c00-8c0d-11eb-83aa-f565eacab40d.png)

* the heading of the results in loops is also really huge (/results/XXXXX.html)
![image](https://user-images.githubusercontent.com/38468371/112202916-3ca9e800-8c12-11eb-9ddd-bc3a5f7e9ea7.png)

* Maybe this is just because I was used to it, but I would put the Status column at the front of the row in the playbook index page.  For me the first column is the eye catcher, and the most important thing is if the playbook was sucessfull or not. And I think the CLI column is not that important, thus I woul put it on the end, just before the Labels columen:
![image](https://user-images.githubusercontent.com/38468371/112200439-86450380-8c0f-11eb-97a9-d142142a2e71.png)
(see hille721/ara@6af62b3a551d0bf0ceb3b9ae3f9e5df0c04c6e2c)

**Comment by dmsimard** [comment]
@hille721 o/ thanks for the feedback

> for me it is really hard to read the labels in the dark theme

Is it about the color contrast or the font size ? What do you think would make it better ?

> due to the sorting option, the Date and Duration column is not aligned in the playbook index page

I know ! I tried to find the right approach to fix it but in the end I ended up with https://github.com/ansible-community/ara/pull/226/commits/c47b0c45565bcb42a48b6d287133aa1f954f957c which made it a bit better but it's not pixel perfect. Suggestions welcome :)

![external-content duckduckgo com](https://user-images.githubusercontent.com/1291204/112207453-8207ee80-8bed-11eb-8b78-6d765d07d5ee.jpg)

> with 90550b3 you made the playbook reports card scrollable and constrained the height, which is great! Becasue the three cards are on the same row and the hosts cards is expanded by default, it would also make sense to expand the Files and Records cards per default as well (currently they are collapsed). At least if the screen is wide enough.

I have no strong opinions on this one, we can do it.

> the heading of the file in the file page (e.g. https://dev.demo.recordsansible.org/files/201724.html) is really huge
> the heading of the results in loops is also really huge (/results/XXXXX.html)

Fixable, thanks for spotting them :)

> Maybe this is just because I was used to it, but I would put the Status column at the front of the row in the playbook index page. For me the first column is the eye catcher, and the most important thing is if the playbook was sucessfull or not. And I think the CLI column is not that important, thus I woul put it on the end, just before the Labels columen

Well, I can't argue too much over this one because it's almost what we had before :)
![Screenshot from 2021-03-23 15-50-36](https://user-images.githubusercontent.com/1291204/112209164-8c2aec80-8bef-11eb-97d7-3df8193ab74c.png)

What I had in mind when re-ordering the columns was a mixture of importance of the item and whether it was fixed-width or not. For example, labels are not fixed-width: there can be any number of labels and they will wrap and take up whatever space they need and so you don't want this column as the first one, it'll be messy and cause the remainder of the columns to be of variable sizes. Paths are also highly variable too but they are kept to a reasonable size by truncating the beginning of the path if they are too long.

And so I thought that fixed-width items belonged better on the left hand side -- they don't take that much space and they also don't cause the rows columns to "zig-zag".

I'm not against this change but I'd like some time to think if there might be a better way.

**Comment by hille721** [comment]
> Is it about the color contrast or the font size ? What do you think would make it better ?

it is about the color contrast. a darker background would it make maybe better. I've played around a bit, makeing the font thiner also helps:
![image](https://user-images.githubusercontent.com/38468371/112217105-90243200-8c22-11eb-8941-3c919c6869c9.png)
here the `font-size` is now 85% and the `font-weight` is 600

> Well, I can't argue too much over this one because it's almost what we had before :)

Yes, as I said, I', used to have the status at the front of the row :)

> And so I thought that fixed-width items belonged better on the left hand side -- they don't take that much space and they also don't cause the rows columns to "zig-zag".

Agree, but the Status item is also a fixed-width item, so no argument to not put the Status column at the front ;)

**Comment by dmsimard** [comment]
re: labels, we're currently using "info", maybe "primary" would be better ?

(from https://bootswatch.com/darkly/)
![Screenshot from 2021-03-23 17-25-11](https://user-images.githubusercontent.com/1291204/112220598-d49cd700-8bfc-11eb-8bdf-15a1f71f40bc.png)

It would also change it for the "light" theme:

![Screenshot from 2021-03-23 17-27-12](https://user-images.githubusercontent.com/1291204/112220817-0e6ddd80-8bfd-11eb-8934-5dbad3bc55e4.png)

**Comment by dmsimard** [comment]
I think it's fine and it also distinguishes labels from "skipped" tasks for which we also used "info":
![Screenshot from 2021-03-23 22-49-33](https://user-images.githubusercontent.com/1291204/112247157-2f98f300-8c2a-11eb-9a36-c53095f014ba.png)

**Comment by dmsimard** [comment]
@hille721 I addressed some of your feedback in https://github.com/ansible-community/ara/pull/230 and provided comparison gifs, please let me know what you think -- it's deployed on https://dev.demo.recordsansible.org/ if you want to look at it.

**Comment by hille721** [comment]
@dmsimard : wow great how fast you implement it! Really nice now, looking forward for the new release :)

**Comment by dmsimard** [comment]
@hille721 I let the new labels simmer a bit and I think they stand out a bit too much compared to the rest of the UI:

![Selection_126](https://user-images.githubusercontent.com/1291204/112403654-3173bc80-8ce5-11eb-916c-d7d1ce1f06ef.png)

The white kind of aggressively jumps at my eyes !

I tried to make them blend in a bit more while having their own "identity" and came up with this:

![Selection_128](https://user-images.githubusercontent.com/1291204/112403871-a0e9ac00-8ce5-11eb-95b7-7d37ee9f31dd.png)

It looks like this in the light theme:

![Selection_129](https://user-images.githubusercontent.com/1291204/112404052-f8881780-8ce5-11eb-9741-028eb9d2e355.png)

What do you think ?

**Comment by hille721** [comment]
> The white kind of aggressively jumps at my eyes !

haha, but yes, that's true.

I like your new approach also, looking less "aggressive" and the readability is also much better (for me :) )

**Comment by hille721** [comment]
@dmsimard: not sure if you already noticed it, but there are a few  bugs in the search of the result card of a playbook page:
![Screenshot_20210326_192928](https://user-images.githubusercontent.com/38468371/112677644-4d549b00-8e6a-11eb-8d1e-2985b89e2d3c.png)

Instead that there is a yellow "Changed" badge in the status row, there is a "Changed: True" Filter
Also all the stati could be in one row? There is massive space on the right. Right now if I am are expanding the filter, on my laptop screen you just see the filter and not more results anymore.

I'm also wondering if it would be possible to make the status badges in the task card clickable, as they are in the hosts card.

**Comment by dmsimard** [comment]
@hille721 o/

> Instead that there is a yellow "Changed" badge in the status row, there is a "Changed: True" Filter

This is because ``changed`` is an attribute, not a status -- but the API muddies that a bit:
https://github.com/ansible-community/ara/blob/e010976578fa87903c6e512d9d7225f0bc16a8e8/ara/api/serializers.py#L27-L36

You might be interested in some additional context of why that is here: https://github.com/ansible-community/ara/issues/150

It's set up like that in part because that's how it's made available through the callback plugin interface when ara picks results up: https://github.com/ansible-community/ara/blob/e010976578fa87903c6e512d9d7225f0bc16a8e8/ara/plugins/callback/ara_default.py#L536

In other words, a task can have ``ok`` for status (if it was successful) while having resulted in a change (or not).

> Also all the stati could be in one row? There is massive space on the right. Right now if I am are expanding the filter, on my laptop screen you just see the filter and not more results anymore.

Good catch, we can fix that :)

> I'm also wondering if it would be possible to make the status badges in the task card clickable, as they are in the hosts card.

Sure, in fact @guillaumevincent just contributed that for the playbook index: https://github.com/ansible-community/ara/pull/235

**Comment by hille721** [comment]
ok, understood the thing with Ok and Changed. Am I right that this is an Ansible speciality rather than an Ara thin?
But idea: instead of only having a "Changed: [] True", we could make a "Changed: [] CHANGED BADGE"
This is just because as I user I am really confused why there is a yellow badge "Changed" in the Status column, but not in the search :)

> Sure, in fact @guillaumevincent just contributed that for the playbook index: #235
yes, that is awesome and also the reason why I was checking als badges if they are clickable :D And yes, that is a bit inconsistent because they are clickable in the hosts card but not the tasks card of the playbook site.

**Comment by dmsimard** [comment]
> Am I right that this is an Ansible speciality rather than an Ara thing?

More or less. Ansible sends us data and we interpret it.

> But idea: instead of only having a "Changed: [] True", we could make a "Changed: [] CHANGED BADGE"
> This is just because as I user I am really confused why there is a yellow badge "Changed" in the Status column, but not in the search :)

Makes sense, we can probably do something like that.

> yes, that is awesome and also the reason why I was checking all badges if they are clickable :D And yes, that is a bit inconsistent because they are clickable in the hosts card but not in the tasks card of the playbook site.

👍

**Comment by dmsimard** [comment]
@hille721 I've added links to result statuses (it was kind of tricky but I think it's OK) and I've adjusted search for ``changed``:
![Screenshot from 2021-04-01 10-05-31](https://user-images.githubusercontent.com/1291204/113306238-dee06480-92d1-11eb-9ed4-e853ca387700.png)

You'll find that in a PR I just sent: https://github.com/ansible-community/ara/pull/244 and it's currently deployed on https://dev.demo.recordsansible.org.

FYI I want to circle back to the host and task ID fields at some point so we can use names instead and I've created an issue so we don't forget it: https://github.com/ansible-community/ara/issues/243. It's not going to make the cut for 1.5.6 but it's something that I would like to see in the future.

**Comment by hille721** [comment]
@dmsimard: thanks, looking good and much more consistent now! Do I see right that you aren't brave enough to go with the dark theme as default? :D But yeah, maybe better :+1: 

But what I realized now, is the rework in the host page, that the cards with the facts is really limited but for it there is now a really large sized tasks size results card. Honestly I do not really like that and in my opinion this is a worsening to the old host site.
You are coming to the hosts site by clicking on "facts"  on the host card on a playbook page. But you are now not really getting what you want, you are getting  a host page with not more than a third of the site with facts, but you are getting all the task results for a host again. But just for that you don't have to go on the host page, you can filter for that directly on the playbook page. 

On following picture you can see how annoying it will be to look for facts. You have hundreds of facts but just this low card:
![image](https://user-images.githubusercontent.com/38468371/113360294-c14ee180-9349-11eb-8661-a1e185314cf4.png)

Btw the font size of the facts could be much smaller as well (in general all the ARA fonts are a bit larger than what you are getting on other sites. For example the font size of the words I'm writing now in GIthub are tiny to the ARA fonts.)

**But I don't won't to open again and again new topics, that was just one more feedback. I would say let's land the new release and further adjustments can be done in next releases** :1st_place_medal: 

> FYI I want to circle back to the host and task ID fields at some point so we can use names instead and I've created an issue so we don't forget it: #243. It's not going to make the cut for 1.5.6 but it's something that I would like to see in the future.

Completely understood that this won't be in the next release! Nevertheless, I looking forward to the possibility to search for hosts, because this will be really a huge improvement, not only on the playbook site, but also on the index page (#218) or directly on a host index page (#206).

**Comment by dmsimard** [comment]
> Do I see right that you aren't brave enough to go with the dark theme as default?

I don't have a particularly strong opinion but I think defaulting to dark lets people know ``there is a dark theme now??`` but on second thought maybe we should default to light and leaving the demo in dark is fine :)

In any case, I'd like to shift the theme selection to be client side (with a dynamic default based on ``prefers-color-scheme`` which was mentioned [here](https://github.com/ansible-community/ara/issues/239#issuecomment-808525438)) and that would be the last chunk that I would like to land before releasing.

> On following picture you can see how annoying it will be to look for facts. You have hundreds of facts but just this low card

Yeah, despite being out of scope for the refactor, the hosts page evolved from showing only host facts to showing task results and then I added the card max-height like for playbook hosts, files and records. I agree that it's sub-optimal and there is surely a better way to present host facts.

I thought about increasing the max-height as a temporary compromise but the current max-height is tuned to allow for at least one task result to be visible at the bottom when the search panel is expanded at 1080p... so increasing max-height could potentially push task results out of view.

I'm thinking perhaps a ``[+]`` button or clicking on the card header would expand to it's full height. It would be convenient on the playbook report page too. In fact, clicking on the card collapses it which doesn't really serve a purpose anymore. Clicking to expand could make sense. Not a release blocker for me but definitely nice to have.

> Btw the font size of the facts could be much smaller as well (in general all the ARA fonts are a bit larger than what you are getting on other sites. For example the font size of the words I'm writing now in GIthub are tiny to the ARA fonts.)

I noticed that too but the funny thing is that we're not doing anything special with the font size, it's like that out of the box from bootstrap though I notice that the font size in darkly is slightly smaller.

**Comment by guillaumevincent** [comment]
@hille721 What is the first thing you would like to see fixed?

**Comment by hille721** [comment]
@dmsimard: possibilty to expand the cards to its full height sounds good!

> I notice that the font size in darkly is slightly smaller.

You ser that to, I thought I was just imagining it. :D


> What is the first thing you would like to see fixed? 

@guillaumevincent : regarding the UI? There is just one thing left: the thing with the Facts card highlighted above.

**Comment by guillaumevincent** [comment]
Ok let me work on the Host Facts card

**Comment by dmsimard** [comment]
o/

The host facts card height restriction was removed and it'll ship that way in 1.5.6 tomorrow.

I'll close this issue now but feel free to open a new one if you find anything in specific.

Thanks !
